admin/cleanup.py

```diff
@@ -230,22 +251,107 @@ def _filter_test_volumes(self, maximum_age, volumes):
                 keep.append(volume)
         return VolumeActions(destroy=destroy, keep=keep)
 
-    def _destroy_cloud_volumes(self, volumes):
-        """
-        Unconditionally and irrevocably destroy all of the given cloud volumes.
-        """
-        for volume in volumes:
-            try:
-                volume.destroy()
-            except:
-                err(None, "Destroying volume.")
+
+@attributes(['lag', 'prefixes'])
+class CleanAcceptanceNodes(object):
+    """
+    :ivar timedelta lag: The age of nodes to destroy.
+    :param prefixes: List of prefixes of nodes to destroy.
+    """
+    def start(self, config):
+        # Get the libcloud drivers corresponding to the acceptance tests.
+        rackspace = get_rackspace_driver(config["rackspace"])
+        ec2 = get_ec2_driver(config["aws"])
+        drivers = [rackspace, ec2]
+
+        # Get the prefixes of the node names, appending the creator from
+        # the config.
+        creator = config['metadata']['creator']
+        prefixes = tuple(map(lambda prefix: prefix + creator, self.prefixes))
+
+        # Find out the cutoff time to use.
+        now = datetime.now(tzutc())
+        cutoff = now - self.lag
+
+        # Get all the nodes from the cloud providers
+        all_nodes = []
+        for driver in drivers:
+            all_nodes += driver.list_nodes()
+
+        # Filter them for running nodes with the right prefix.
+        test_nodes = [
+            node for node
+            in all_nodes
+            if node.name.startswith(prefixes)
+            # Also, terminated nodes that still show up.  An OpenStack bug
+            # causes these to hang around sometimes.  They're not billed in
+            # this state but they do count towards RAM quotas.  Quoth Rackspace
+            # support:
+            #
+            # > The complete fix for this issue is expected in the next
+            # > Openstack iteration (mid August).  Until then what can be done
+            # > is just to issue another delete against the same node.  The
+            # > servers are only billed when they are in Active (green) status,
+            # > so the deleted nodes are not billed.
+            #
+            # So consider any nodes in that state as potential destruction
+            # targets.
+            and node.state in (NodeState.RUNNING, NodeState.TERMINATED)
```

Comment:
:question: Still a bit of a plug for abstracting this behind cloud-specific integrations. That way you could hide the ugly comment above in some specific rackspace implementation, rather than also doing it for EC2 (I think you are also doing it for EC2 here, which I guess isn't ideal?).